### PR TITLE
SOLR-17106: LBSolrClient - Make zombie ping checks configurable

### PR DESCRIPTION
**Pull request by aparnasuresh85**
https://issues.apache.org/jira/browse/SOLR-17106

# Description

Please provide a short description of the changes you're making with this pull request.
The issue involves scalability challenges in SolrJ's LBSolrClient when a pod with numerous cores experiences connectivity problems. The "zombie" ping checking mechanism, operating on a core basis, becomes a bottleneck during distributed search on a massive multi shard collection. Threads attempting to reach unhealthy cores contribute to a high computational load, causing performance issues. 

LBSolrClient could be configured to disable zombie "ping" checks, but retain zombie tracking. Once a replica/endpoint is identified as a zombie, it could be held in zombie state for X seconds, before being released - hoping that by this timeframe ZK would be updated to mark this endpoint DOWN or the pod is back up and CloudSolrClient would avoid querying it. In any event, only 1 failed query would be needed to send the server back to zombie state.

Dev list thread: https://lists.apache.org/thread/f0zfmpg0t48xrtppyfsmfc5ltzsq2qqh

# Solution

Please provide a short description of the approach taken to implement your solution.

Modified the code to make zombie ping checking configurable through a new property named "enableZombiePingChecks." When set to true, the system is designed to periodically send zombie ping requests every zombiePingIntervalMillis. Conversely, setting it to false disables zombie pings and instead monitors the "zombieServers" every zombieStateMonitoringIntervalMillis. This monitoring identifies replicas or endpoints that have spent at least minZombieReleaseTimeMillis as zombies, subsequently releasing them from the zombie state and adding them back to the alive set.

The default configuration for zombiePingIntervalMillis remains at 1 minute, while zombieStateMonitoringIntervalMillis is currently set to 5 seconds, and minZombieReleaseTimeMillis is configured at 10 seconds.

Corresponding tests in TestLBHttp2SolrClient and TestLBHttpSolrClient have been adjusted accordingly.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.
Slight refactoring in TestLBHttp2SolrClient and TestLBHttpSolrClient, added a new test testReliabilityWithoutZombiePingChecks in TestLBHttp2SolrClient. All tests pass.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttp2SolrClient.java [hunk 1]
```diff
@@ -281,6 +298,37 @@ public LBHttp2SolrClient.Builder setAliveCheckInterval(int aliveCheckInterval, T
       return this;
     }
 
+    /**
+     * Positive values can be used to ensure that the liveness checks for a given URL will only be
+     * run every "Nth" iteration of the setAliveCheckInterval
+     *
+     * @param aliveCheckSkipIters, number of iterations to skip, an int value *
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttp2SolrClient.java:305]: did this mean to have a trailing `*`?

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttp2SolrClient.java [hunk 2]
```diff
@@ -281,6 +298,37 @@ public LBHttp2SolrClient.Builder setAliveCheckInterval(int aliveCheckInterval, T
...
+     * @see #setAliveCheckInterval(int, TimeUnit)
+     * @see #setAliveCheckQuery(SolrQuery)
+     */
+    public LBHttp2SolrClient.Builder setAliveCheckSkipIters(int aliveCheckSkipIters) {
+      if (aliveCheckSkipIters < 0) {
+        throw new IllegalArgumentException(
+            ("Alive Check Skip Iterations must be positive, specified value = "
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttp2SolrClient.java:312]: do we need the doubled `(` here?   Maybe for string parsing?

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttpSolrClient.java [hunk 3]
```diff
@@ -266,19 +280,49 @@ public Builder withBaseSolrUrls(String... solrUrls) {
...
+            "Alive check interval must be positive, specified value = " + aliveCheckInterval);
       }
       this.aliveCheckInterval = aliveCheckInterval;
       return this;
     }
 
+    /**
+     * Positive values can be used to ensure that the liveness checks for a given URL will only be
+     * run every "Nth" iteration of the setAliveCheckInterval
+     * @param aliveCheckSkipIters, number of iterations to skip, an int value *
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttpSolrClient.java:300]: again the trailing `*`

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttpSolrClient.java [hunk 4]
```diff
@@ -266,19 +280,49 @@ public Builder withBaseSolrUrls(String... solrUrls) {
...
+     * @see #setAliveCheckInterval(int)
+     * @see #setAliveCheckQuery(SolrQuery)
+     */
+    public Builder setAliveCheckSkipIters(int aliveCheckSkipIters) {
+      if (aliveCheckSkipIters < 0) {
+        throw new IllegalArgumentException(
+            ("Alive Check Skip Iterations must be positive, specified value = "
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttpSolrClient.java:307]: and the doubled (

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java [hunk 5]
```diff
@@ -79,24 +81,27 @@ public abstract class LBSolrClient extends SolrClient {
 
   protected long aliveCheckIntervalMillis =
       TimeUnit.MILLISECONDS.convert(60, TimeUnit.SECONDS); // 1 minute between checks
+  protected int aliveCheckSkipIters = 0;
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java:84]: Do we use the short hand `Iters` in lots of places?   If not, I like the full `Iteration` as just one less thing to read.

Comment (epugh) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java:84]: maybe just a personal thing though.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java [hunk 6]
```diff
@@ -480,35 +488,70 @@ public RequestWriter getRequestWriter() {
     return requestWriter;
   }
 
+  private boolean isServerAlive(ServerWrapper zombieServer)
+      throws SolrServerException, IOException {
+    if (null == aliveCheckQuery) {
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java:493]: I think we have normally a check to see if we are in debug mode before calling the `.debug`

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java [hunk 7]
```diff
@@ -480,35 +488,70 @@ public RequestWriter getRequestWriter() {
...
+      log.debug("Assuming success because aliveCheckQuery is null");
+      return true;
+    }
+    log.debug("Running ping check on server " + zombieServer.getBaseUrl());
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java:497]: likewise on the check?

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java [hunk 8]
```diff
@@ -480,35 +488,70 @@ public RequestWriter getRequestWriter() {
...
-          wrapper.failedPings = 0;
-          if (wrapper.standard) {
-            addToAlive(wrapper);
-          }
-        } else {
-          // something else already moved the server from zombie to alive
-        }
+      // push back on liveness checks only every Nth iteration
+      if (zombieServer.skipAliveCheckIters > 0) {
+        log.debug(
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java:535]: the ifDebugEnabled check

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java [hunk 9]
```diff
@@ -480,35 +488,70 @@ public RequestWriter getRequestWriter() {
...
-    } catch (Exception e) {
-      // Expected. The server is still down.
-      zombieServer.failedPings++;
 
-      // If the server doesn't belong in the standard set belonging to this load balancer
-      // then simply drop it after a certain number of failed pings.
-      if (!zombieServer.standard && zombieServer.failedPings >= NONSTANDARD_PING_LIMIT) {
-        zombieServers.remove(zombieServer.getBaseUrl());
+      if (isServerAlive(zombieServer)) {
+        log.debug("Successfully pinged server " + zombieServer.getBaseUrl() + ", marking it alive");
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java:545]: and the check

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java [hunk 10]
```diff
@@ -480,35 +488,70 @@ public RequestWriter getRequestWriter() {
...
+        handleServerBackUp(zombieServer);
       }
+
+    } catch (Exception e) {
+      log.debug(
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java:550]: and the check!   I think the whole check thing is werid that it's not just part of log.debug!

File: solr/solrj/src/test/org/apache/solr/client/solrj/TestLBHttp2SolrClient.java [hunk 11]
```diff
@@ -205,16 +202,31 @@ public void testTwoServers() throws Exception {
     }
   }
 
-  public void testReliability() throws Exception {
+  @LogLevel("org.apache.solr.client.solrj.impl.LBSolrClient=DEBUG")
```
Comment (epugh) [inline solr/solrj/src/test/org/apache/solr/client/solrj/TestLBHttp2SolrClient.java:205]: is this needed to make the test pass, or from working on the code?  the @logLevel?

Comment (epugh) [reply solr/solrj/src/test/org/apache/solr/client/solrj/TestLBHttp2SolrClient.java:205]: ah, I think I see why you have it from looking lower down in the LogReliablityTestSetup

**Review by epugh — CHANGES_REQUESTED** [review]
We need to update the merge conflicts, review some of the formatting etc.  And lastly, to use the right method of looking up settings.  I think @dsmiley has already approved the direction, which makes me more comfortable about merging it.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java (hunk 9 above)
Comment (dsmiley) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java:545]: Always use "{}" templates in SLF4J

**Pull request closed without merging**